File: tests/fixtures/fsharp/Sample.fs
namespace App.Service

open System
open App.Store.Repo

/// A user record.
type User =
    { Id: int
      Name: string }

/// Account status.
type Status =
    | Active
    | Suspended of reason: string
    | Closed

/// Things that can be stored.
type IStore =
    abstract member Get: int -> User option
    abstract member Put: User -> unit

/// In-memory store.
type MemoryStore(capacity: int) =
    inherit BaseStore()
    let mutable count = 0

    /// Current item count.
    member this.Count = count

    member this.Add(u: User) =
        count <- count + 1
        u

    interface IStore with
        member this.Get id = None
        member this.Put u = ()

module Users =
    /// Find a user by id.
    let findUser (table: Map<int, User>) uid = Map.tryFind uid table

    let rec normalize (s: string) = s.ToLower()

    let renderUser (u: User) = normalize u.Name

    let maxRetries = 3

    let apiToken = Environment.GetEnvironmentVariable "API_TOKEN"

    [<Test>]
    let ``normalize is idempotent`` () =
        normalize "AB" = normalize (normalize "AB")

    [<Fact>]
    let testRender () =
        let store = MemoryStore(4)
        store.Add({ Id = 1; Name = "a" }) |> ignore

File: tests/fixtures/ocaml/sample.ml
(** Sample module for extractor tests. *)

open Core
open Store.Repo
include Utils

(** A user record. *)
type user = {
  id : int;
  name : string;
}

(** Account status. *)
type status =
  | Active
  | Suspended of string
  | Closed

type user_table = (int, user) Hashtbl.t

module type STORE = sig
  val get : int -> user option
  val put : user -> unit
end

module Store = struct
  (** In-memory table. *)
  let table : user_table = Hashtbl.create 16

  let get id = Hashtbl.find_opt table id

  let put u = Hashtbl.replace table u.id u
end

class counter init = object
  val mutable n = init
  method bump () = n <- n + 1
  method value = n
end

(** Find a user by id. *)
let find_user table uid =
  Hashtbl.find_opt table uid

let rec render_user u = normalize u.name
and normalize s = String.lowercase_ascii s

let max_retries = 3

let api_token = Sys.getenv "API_TOKEN"

let%test "normalize is idempotent" = normalize "AB" = normalize (normalize "AB")
